driver/src/orderbook/streamed/state.rs

```diff
@@ -84,97 +85,204 @@ impl State {
+    fn apply_withdraw_request(&mut self, event: &WithdrawRequest) -> Result<(), Error> {
         let balance = self.balances.entry((event.user, event.token)).or_default();
-        balance.pending_withdraw = Some(Flux {
-            amount: event.amount,
-            batch_id: event.batch_id,
-        });
+        balance.withdraw_request(event.amount, event.batch_id);
         Ok(())
     }
 
-    fn withdraw(&mut self, event: &Withdraw, block_batch_id: BatchId) -> Result<(), Error> {
+    fn apply_withdraw(&mut self, event: &Withdraw, block_batch_id: BatchId) -> Result<(), Error> {
         let balance = self.balances.entry((event.user, event.token)).or_default();
         balance.withdraw(event.amount, block_batch_id)
     }
 
-    fn token_listing(&mut self, event: &TokenListing) -> Result<(), Error> {
+    fn apply_token_listing(&mut self, event: &TokenListing) -> Result<(), Error> {
         self.tokens.0.push((event.id, event.token));
         Ok(())
     }
 
-    fn order_placement(&mut self, event: &OrderPlacement) -> Result<(), Error> {
-        let order = Order {
-            buy_token: event.buy_token,
-            sell_token: event.sell_token,
-            valid_from: event.valid_from,
-            valid_until: event.valid_until,
-            price_numerator: event.price_numerator,
-            price_denominator: event.price_denominator,
-            used_amount: 0,
-        };
+    fn apply_order_placement(&mut self, event: &OrderPlacement) -> Result<(), Error> {
+        let order = Order::new(
+            event.buy_token,
+            event.sell_token,
+            event.valid_from,
+            event.valid_until,
+            event.price_numerator,
+            event.price_denominator,
+        );
         match self.orders.entry((event.owner, event.index)) {
             Entry::Vacant(entry) => entry.insert(order),
             Entry::Occupied(_) => return Err(Error::OrderAlreadyExists),
         };
         Ok(())
     }
 
-    fn order_cancellation(
+    fn apply_order_cancellation(
         &mut self,
         event: &OrderCancellation,
         block_batch_id: BatchId,
     ) -> Result<(), Error> {
         let order = self
             .orders
             .get_mut(&(event.owner, event.id))
-            .ok_or(Error::UnknownOrder)?;
+            .ok_or(Error::UnknownOrder(event.id))?;
         order.valid_until = block_batch_id - 1;
         Ok(())
     }
 
-    fn order_deletion(&mut self, event: &OrderDeletion) -> Result<(), Error> {
+    fn apply_order_deletion(
+        &mut self,
+        event: &OrderDeletion,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        if let Some(order) = self.orders.get(&(event.owner, event.id)) {
+            if order.is_valid_in_batch(block_batch_id - 1) {
+                return Err(Error::DeletingValidOrder);
+            } else {
+                self.orders.remove(&(event.owner, event.id));
+            }
+        }
```

Comment:
Nit: the comment below is outdated now. We could still unconditionally remove the order if you prefer.